### DISCUSSION TITLE
How can I conditionally navigate to child routes in a navigator?

### DISCUSSION THREAD

**Discussion by eliw00d** [react-navigation/react-navigation#10439] [Q&A] [answered]
Most of the time, the initial route is sufficient, but there are times when I need to navigate to a specific child based on certain conditions. With react-router-dom, this is easy, I just do:
```jsx
useEffect(() => {
  if (conditionB) {
    history.replace(pathB)
  } else if (conditionC) {
    history.replace(pathC)
  }
}, [history])

...

return (
  <Switch>
    <Route path={pathA}>
      <ComponentA />
    </Route>
    <Route path={pathB}>
      <ComponentB />
    </Route>
    <Route path={pathC}>
      <ComponentC />
    </Route>
  </Switch>
)
```

However, with react-navigation, if I do:
```jsx
useEffect(() => {
  if (conditionB) {
    navigation.replace('Parent', { screen: pathB })
  } else if (conditionC) {
    navigation.replace('Parent', { screen: pathC })
  }
}, [navigation])

...

return (
  <Stack.Navigator>
    <Stack.Screen name={pathA} component={ComponentA} />
    <Stack.Screen name={pathB} component={ComponentB} />
    <Stack.Screen name={pathC} component={ComponentC} />
  </Stack.Navigator>
)
```

the effect runs in an infinite loop. I also tried `useLayoutEffect`, `useFocusEffect`, and even moving the useEffect into ComponentA since it is the initial route but to no avail.

What is the best approach for react-navigation to accomplish the same thing? I can't find anything in the documentation for this.

**Comment by raajnadar** [comment, marked as answer]
why not make the initial route name dynamic??

```initialRouteName={test ? 'Home' : 'Settings'}```

https://reactnavigation.org/docs/stack-navigator#initialroutename

**Reply by eliw00d** [reply]
I feel like I tried that as well, but I will double check just to be sure and get back to you.

**Reply by eliw00d** [reply]
I must have been in a weird state. That does work. It would be nice for it to work in a similar way to react-router, but this is fine for now.

**Reply by raajnadar** [reply]
The architecture for react js web & react-native is different. In react-navigation, the previous screen is still mounted (like a stack) on the memory which is the reason you get an infinite loop.

Another clean solution is to check for the condition in the HOC or in the previous screen and pass the screen name as props and pass the same to the initialRouteName props of the navigator.

**Reply by eliw00d** [reply]
I get that they are different, I just figured replacing would remove the current screen from the stack and replace it with the specified screen. Maybe I misunderstood how replace works with this library. I'll keep the other solutions in mind but the initialRouteName seems good enough. Thanks again!
